flow/Platform.cpp

```diff
@@ -1835,23 +1944,24 @@ std::string abspath( std::string const& filename ) {
 #endif
 }
 
-std::string basename( std::string const& filename ) {
-	auto abs = abspath(filename);
+std::string parentDirectory( std::string const& path, bool resolveLinks, bool mustExist ) {
+	auto abs = abspath(path, resolveLinks, mustExist);
 	size_t sep = abs.find_last_of( CANONICAL_PATH_SEPARATOR );
-	if (sep == std::string::npos) return filename;
-	return abs.substr(sep+1);
+	if (sep == std::string::npos) {
+		Error e = platform_error();
+		TraceEvent(SevWarnAlways, "GetParentDirectory")
+			.detail("File", path)
+			.error(e);
+		throw e;
+	}
+	return abs.substr(0, sep + 1);
 }
 
-std::string parentDirectory( std::string const& filename ) {
+std::string basename( std::string const& filename ) {
```

Comment:
Should this have a resolve symlinks flag too?

Reply:
Perhaps, and perhaps it should default to false.  Currently, all usages of basename() are passing what they believe to be a path to a file and just want to get the filename part from the end of the path.  As-written (and this behavior has not been changed by this PR) if the filename argument was actually a path to a symbolic link which then resolves to some other file or directory, basename() would return the name of the final resolved file or directory.  I actually think this is not what any of the use cases want.

Reply:
Interesting. This was my thought as well, namely that I would expect the caller to usually not want to resolve links. I guess that's a problem for another time though.